### PR TITLE
Add VALID UNTIL clause for users

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow specifying the expiration date and, optionally, the time for user credentials with `VALID UNTIL datetime` clause.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 0d2b9fd0e5bb5ba50317260e7c3bcc2fea6f420c with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51261/0d2b9fd0e5bb5ba50317260e7c3bcc2fea6f420c/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Access/IAccessStorage.cpp [hunk 1]
```diff
@@ -514,6 +514,14 @@ bool IAccessStorage::areCredentialsValid(
     if (credentials.getUserName() != user.getName())
         return false;
 
+    if (user.valid_until)
+    {
+        const time_t now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+
+        if (now > user.valid_until)
+            return false;
```
Comment (pufit) [inline src/Access/IAccessStorage.cpp:522]: This will throw `ErrorCodes::WRONG_PASSWORD`, which is probably not the correct error

Comment (pufit) [reply src/Access/IAccessStorage.cpp:522]: Also, shouldn't we remove expired users after some time?

File: tests/integration/test_user_valid_until/test.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    node.query("ALTER USER user_basic VALID UNTIL '06/11/2040 08:03:20 Z+3'")
+
+    assert (
+        node.query("SHOW CREATE USER user_basic")
+        == "CREATE USER user_basic VALID UNTIL \\'2040-11-06 05:03:20\\'\n"
+    )
+    assert node.query("SELECT 1", user="user_basic") == "1\n"
+
+    # 3. With invalid VALID UNTIL
+    node.query("ALTER USER user_basic VALID UNTIL '06/11/2010 08:03:20 Z+3'")
```
Comment (vitlibar) [inline tests/integration/test_user_valid_until/test.py:36]: What is invalid here?

**Comment by alexey-milovidov** [comment]
Please continue.

File: src/Access/IAccessStorage.cpp (hunk 1 above)
Comment (evillique) [reply src/Access/IAccessStorage.cpp:522]: The `VALID UNTIL` clause is a way to set the expiration date for the credentials of the user, similar to `VALID UNTIL` in PostgreSQL. So after the expiration date has passed not the user but the password becomes invalid, so the `ErrorCodes::WRONG_PASSWORD` error sounds valid and we also don't need to necessarily remove those users.

**Review by vitlibar — APPROVED** [review]

**Comment by vitlibar** [comment]
I will fix the problem with [test_grpc_procol](https://s3.amazonaws.com/clickhouse-test-reports/51261/0998b0bbc740f26fccd0d01f23014b40c75b6359/integration_tests__asan__analyzer__[4_6].html) soon.

**Pull request merged**